File: src/main/java/io/karma/pda/api/common/app/component/DefaultContainer.java
/*
 * Copyright (C) 2024 Karma Krafts & associates
 */

package io.karma.pda.api.common.app.component;

import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.Objects;
import java.util.UUID;

/**
 * @author Alexander Hinze
 * @since 08/02/2024
 */
public class DefaultContainer extends AbstractComponent implements Container {
    protected final LinkedHashMap<UUID, Component> children = new LinkedHashMap<>();

    public DefaultContainer(final ComponentType<?> type, final UUID id) {
        super(type, id);
    }

    @Override
    public @Nullable Component findChild(final UUID id) {
        return children.get(id);
    }

    @Override
    public Collection<Component> getChildren() {
        return children.values();
    }

    @Override
    public void addChild(final Component child) {
        children.put(child.getId(), child);
        child.setParent(this);
        flexNode.addChild(child.getFlexNode());
    }

    @Override
    public void removeChild(final Component child) {
        children.remove(child.getId());
        child.setParent(null);
        flexNode.removeChild(child.getFlexNode());
    }

    @Override
    public boolean equals(final Object obj) {
        if (!(obj instanceof Container container)) {
            return false;
        }
        for (final var childId : children.keySet()) {
            if (container.findChild(childId) != null) {
                continue;
            }
            return false;
        }
        return id.equals(container.getId());
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, children);
    }
}

File: src/composable/kotlin/io/karma/pda/composition/component/DefaultContainer.kt
/*
 * Copyright (C) 2024 Karma Krafts & associates
 */

@file:JvmName("ComponentExtensionsKt")
@file:JvmMultifileClass

package io.karma.pda.composition.component

import io.karma.pda.api.app.component.Container
import io.karma.pda.composition.Composable
import io.karma.pda.composition.Composer
import io.karma.pda.composition.LayoutProps
import io.karma.pda.foundation.component.DefaultComponents
import io.karma.pda.foundation.component.DefaultContainer

/**
 * @author Alexander Hinze
 * @since 07/06/2024
 */

@Composable
inline fun <C : Container> Composer<C>.DefaultContainer(
    crossinline layoutProps: LayoutProps,
    localName: String? = null,
    props: @Composable DefaultContainer.() -> Unit = {},
    content: @Composable Composer<DefaultContainer>.() -> Unit
) = container(
    type = DefaultComponents.CONTAINER,
    layoutProps = layoutProps,
    localName = localName,
    props = props,
    content = content
)